<a href="https://colab.research.google.com/github/DiahKurnillah/Praktikum-Komputasi-Biomedis/blob/main/KombioPrak_ChapIX.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Chapter 9. Numerical Integration: Trapezoid & Simpson’s Rule**
---
**Tujuan**: Untuk menentukan integrasi numerik menggunakan metode Trapezium dan Simpson 1/3

**Literarure Review**
Perhitungan integrasi analitik dilakukan dengan menggunakan titik-titik diskrit yang dibagi menjadi beberapa bagian yang disebut dengan "segmen". Ada beberapa metode yang menggunakan metode segmen, seperti metode trapesium dan metode persegi panjang.

Luas dari satu persegi panjang adalah
$$\int_{x_0}^{x_1} f(x) \,dx \approx h f(x_0) $$
$$\int_{x_0}^{x_1} f(x) \,dx \approx h f(x_1) $$
$$\int_{x_0}^{x_1} f(x) \,dx \approx \frac{h}{2} [f(x_0) + f(x_1)] $$

persamaan-persamaan tersebut dapat disederhanakan menjadi
$$\int_a^b f(x) \,dx \approx \frac{h}{2} (f_0 + 2f_1 + 2f_2 + \ldots + 2f_{n-1} + f_n) = \frac{h}{2} (f_0 + 2\sum_{i=1}^{n-1} f_i + f_n)$$

Metode trapesium hampir sama dengan metode persegi panjang. Bentuk yang digunakan untuk membagi area di bawah kurva adalah trapesium. Dengan menggunakan metode ini, diharapkan dapat meminimalisir kesalahan yang dihasilkan oleh bentuk persegi panjang pada metode rectangular.

$$ \int_a^b f(x) \,dx \approx \int_{x_0}^{x_1} f(x) \,dx + \int_{x_1}^{x_2} f(x) \,dx + \ldots + \int_{x_{n-1}}^{x_n} f(x) \,dx $$
$$ \approx \frac{h}{2} [f(x_0) + f(x_1)] + \frac{h}{2} [f(x_1) + f(x_2)] + \ldots + \frac{h}{2} [f(x_{n-1}) + f(x_1)] $$
$$\approx \frac{h}{2}[f(x_0) + 2f(x_1) + 2f(x_2) + \ldots + 2f(x_{n-1}) + f(x_n)] $$
$$\approx \frac{h}{2}(f_0 + 2\sum_{i=1}^{n-1} f_1 + f_n) $$

Selain metode "segmen", ada metode lain untuk menghitung integrasi suatu fungsi, seperti metode Newton Cotes. Metode ini menggunakan interpolasi polinomial, seperti metode trapesium, metode Simpson 1/3, dan metode Simpson 3/8.
Metode Simpson 1/3 membutuhkan minimal 3 titik untuk menentukan hampiran integrasi suatu fungsi, sebagai contoh (0, f(0)), (h, f(h)), dan (2h, f(2h))

$$I_{\text{tot}} = \int_a^b f(x) \,dx \approx \int_{x_0}^{x_2} f(x) \,dx + \int_{x_2}^{x_4} f(x) \,dx + \ldots + \int_{x_{n-2}}^{x_{2n}} f(x) \,dx $$
$$\approx \frac{h}{3}(f_0 + 4\sum_{i=1,3,5}^{n-1} f_i + 2\sum_{i=2,4,6}^{n-2} f_i + f_n)$$


# Preliminary Task
Jika terdapat fungsi sebagai berikut:
$f(x) = x^2 \cos(x^2), \quad 1.5 \leq x \leq 2.5$
dan $( h = 0.1 )$.

Hitunglah integrasi berikut dengan menggunakan metode trapezium dan Simpson 1/3:
\$\int_{0}^{2} f(x) \,dx$

In [ ]:
import numpy as np

# Fungsi
def f(x):
    return x**2 * np.cos(x**2)

# Batasan
a = 0
b = 2
h = 0.1
n = int((b - a) / h)

# Metode Trapezium
integral_trapezium = h/2 * (f(a) + 2*np.sum(f(np.linspace(a+h, b-h, n-1))) + f(b))

# Metode Simpson 1/3
integral_simpson13 = h/3 * (f(a) + 4*np.sum(f(np.linspace(a+h, b-h, n//2))) +
                            2*np.sum(f(np.linspace(a+2*h, b-2*h, n//2-1))) + f(b))

print("Hasil Integrasi (Metode Trapezium):", integral_trapezium)
print("Hasil Integrasi (Metode Simpson 1/3):", integral_simpson13)


Hasil Integrasi (Metode Trapezium): -1.1512833926545998
Hasil Integrasi (Metode Simpson 1/3): -1.1591682849187996


Dalam kode di atas, kita mengimplementasikan metode numerik untuk menghitung integral suatu fungsi menggunakan metode trapesium dan metode Simpson 1/3. Kita ingin mengestimasi integral fungsi $f(x)= x^2  cos⁡(x^2 )$   dari 0 hingga 2.
Pertama-tama, kita mendefinisikan fungsi f(x) yang akan diintegrasikan. Selanjutnya, kita menentukan batas integral, yaitu a (0) dan b (2), serta lebar langkah h (0.1) dan jumlah langkah n berdasarkan rumus n=(b-a)/h
Metode trapesium digunakan untuk menghitung integral. Langkah pertama dalam metode ini adalah menghitung nilai integral pada ujung interval (a dan b) dan kemudian menambahkan dua kali lipat jumlah nilai integral pada titik-titik dalam interval dengan langkah h. Hasil akhirnya dikalikan dengan h/2. Ini menciptakan area segitiga di setiap interval yang, ketika diakumulasikan, memberikan perkiraan integral total.

Metode Simpson 1/3 juga digunakan untuk menghitung integral. Pada metode ini, kita membagi interval menjadi segmen-segmen yang lebih kecil dan menggunakan polinomial orde dua (parabola) untuk memperkirakan nilai integral di setiap segmen. Kita kemudian menggabungkan hasilnya dengan bobot khusus, dengan bobot 1 pada ujung interval dan bobot 4 pada titik-titik ganjil di dalam interval, serta bobot 2 pada titik-titik genap di dalam interval.
Setelah menghitung integral menggunakan kedua metode, hasilnya dicetak. Metode trapesium memberikan estimasi integral dengan menggunakan trapesium sebagai bentuk pendekatan area di bawah kurva, sementara metode Simpson 1/3 menggunakan pendekatan parabola.

Dilakukan pula perhitungan secara analitik, didapat
Metode Trapesium
$$∫_0^2f(x)dx=\frac{h}{2}(f_0+2f_1+⋯+2f_19+f_20)$$
$$∫_0^2f(x)dx=\frac{0.1}{2} (-23.0126)$$
$$∫_0^2f(x)dx=-1.151$$

Metode Simpson 1/3
$$∫_0^2f(x)dx=\frac{h}{3}(f_0+4f_1+2f_2+⋯+2f_18+4f_19+f_20)$$
$$∫_0^2f(x)dx=\frac{0.1}{3} (-34.775)$$
$$∫_0^2f(x)dx=-1.159$$
Maka, berdasarkan penghitungan, hasil integral kedua metode Trapesium dan simphson 1/3 sama dengan hasil analitik.


# Task 1
Seorang sky-driver jatuh dari pesawat dengan kecepatan seperti
yang ditunjukkan pada persamaan di bawah ini.
$$v(t) = gm/c(1-e^-(c/m)t$$

Hitung seberapa jauh sky diver jatuh setelah 12 detik dengan beberapa
metode integrasi numerik dan bandingkan hasil masing-masing metode satu
sama lain. Tentukan kesalahan dibandingkan dengan hasil analisis.
Metode mana yang paling akurat? Jelaskan!

v = kecepatan (m/s)

g = percepatan gravitasi = 9,8 m/s2

m = massa penyelam langit = 78,5 kg

c = koefisien resistif udara = 12,5 kg/s


In [ ]:
import math

def integral(t):
    g = 9.8
    m = 78.5
    c = 12.5
    return (g * m / c) * (1 - math.exp(-(c / m) * t))

def eksak(t):
    return ((61544 * math.exp(-0.159 * t)) / 159) + 61.544 * t

a = 0
b = 12
h = 0.01

# Integral Exact
exact = eksak(b) - eksak(a)
print('Integral Exact: ', exact)
print('----------------------------------------------------------')

# Metode Trapezoid
fa = integral(a)
fb = integral(b)
jum = 0

while a < b - h:
    a = a + h
    fx = integral(a)
    jum = jum + fx

hasil_trapezoid = (h / 2) * (fa + 2 * jum + fb)
error_trapezoid = math.fabs(hasil_trapezoid - exact)
print('Integral Trapezoid: ', hasil_trapezoid)
print('Error Trapezoid: ', error_trapezoid)

# Metode Simpson 1/3
n = 0
a = 0
genap = 0
ganjil = 0

while a < b - h:
    a = a + h
    n = n + 1
    m = n % 2
    if m == 0:
        fx = integral(a)
        genap = genap + fx
    else:
        fx = integral(a)
        ganjil = ganjil + fx

hasil_simpson = (h / 3) * (fa + (4 * ganjil) + (2 * genap) + fb)
error_simpson = math.fabs(hasil_simpson - exact)
print('----------------------------------------------------------')
print('Integral Simpson 1/3: ', hasil_simpson)
print('Error Simpson 1/3: ', error_simpson)


Integral Exact:  408.8909211970591
----------------------------------------------------------
Integral Trapezoid:  409.74114536350385
Error Trapezoid:  0.8502241664447752
----------------------------------------------------------
Integral Simpson 1/3:  409.5664213344649
Error Simpson 1/3:  0.6755001374058338


Berdasarkan hasil yang tertera, dapat diamati bahwa nilai eksak untuk jarak yang ditempuh oleh skydiver setelah terjun selama 12 detik adalah sekitar 408.8909211970591 meter. Saat menggunakan metode trapesium, hasil yang diperoleh sebesar 409.74114536350385 dengan nilai error sekitar 0.8502241664447752. Sementara itu, dengan menerapkan metode Simpson 1/3, diperoleh hasil sekitar 409.5664213344649 dengan nilai error sekitar 0.6755001374058338. Perbandingan antara kedua metode menunjukkan bahwa metode Simpson 1/3 memiliki nilai error yang lebih kecil dibandingkan dengan metode trapesium, menunjukkan bahwa metode Simpson 1/3 lebih akurat dalam pendekatan perhitungan jarak yang ditempuh oleh skydiver selama 12 detik

# Task 2
**Infus IV**
Kantong silinder dengan jari-jari 10 cm dan tinggi 50 cm berisi larutan garam (0,9% NaCl) yang diberikan kepada pasien sebagai infus intravena (IV). Metode yang digunakan untuk memasukkan larutan garam ke dalam aliran darah pasien dalam contoh ini adalah infus gravitasi. Larutan mengalir dari kantong ke bawah di bawah pengaruh gravitasi melalui lubang berjari-jari 1 mm di bagian bawah kantong. Jika satu-satunya hambatan yang ditawarkan oleh sistem aliran adalah hambatan viskositas melalui pipa, tentukan waktu yang diperlukan untuk mengosongkan kantong hingga 90%. Panjang pipa adalah 36 (91,44 cm) dan ID (diameter dalam) adalah 1 mm. Viskositas larutan garam = 0,01 Poise, dan massa jenis = 1 g/cm3.
Misalkan L adalah panjang pipa, d diameter pipa, dan R jari-jari kantong silinder. Maka, L = 91,44 cm; d = 0,1 cm; R = 10 cm.

**Keseimbangan energi mekanik**

Keseimbangan energi mekanik diberikan oleh persamaan Bernoulli yang dikoreksi untuk kehilangan energi mekanik akibat gesekan fluida. Energi mekanik pada tingkat cairan di dalam kantong sama dengan energi mekanik fluida yang mengalir keluar dari tabung dikurangi dengan kehilangan gesekan. Kami mengabaikan kehilangan gesekan akibat kontraksi tiba-tiba dari luas penampang aliran di pintu masuk tabung. Ketinggian di mana cairan infus memasuki pasien dianggap sebagai tingkat datum. Ketinggian cairan di dalam kantong adalah z cm terhadap dasar kantong, dan (z + L) cm terhadap level datum

Penurunan tekanan dalam pipa akibat gesekan dinding diberikan oleh persamaan Hagen-Poiseuille:
$$∆p=32Luμ/d^2 =2926.08u$$
Persamaan Bernoulli untuk sistem ini adalah sebagai berikut:
$$g(z+L)=u^2/2+∆p/ρ+32Luμ/(d^2 ρ)$$
Di mana Δ𝑝/𝜌 adalah istilah untuk kerugian mekanis akibat gesekan fluida dan memiliki satuan energi per massa. Neraca massa kondisi tunak Tidak ada reaksi, akumulasi, atau penipisan di dalam sistem, dan kami mengabaikan aliran tidak stabil awal saat memulai infus. Laju aliran massa di dalam kantong sama dengan laju aliran massa di dalam pipa:
$$-ρπR^2  dz/dt=ρ π/4 d^2 u$$
Pada pengaturan ulang, kami mendapatkan
$$dt=-(4R^2)/(d^2 u) dz$$
Kita dapat mengekspresikan u dalam bentuk z dengan menyelesaikan Persamaan untuk u. Membiarkan$ a = 64Lμ/d2 ρ$
$$u=(-a+√(a^2+8g(z+L)))/2$$
Mengganti hal di atas ke dalam neraca massa kondisi tunak
$$dt=-(8R^2)/((-a+√(a^2+8g(z+L) )) d^2 ) dz$$
Mengintegrasikan persamaan diferensial Kita ingin mengintegrasikan z dari 50 cm ke 5 cm. Dengan mengintegrasikan sisi kiri dari 0 ke t, kita memperoleh:
$$t=∫_{50}^5-\frac{8R^2}{(-a+√(a^2+8g(z+L) ) d^2 } dz$$
$$=\frac{8R^2}{d^2}  ∫_{50}^5\frac{1}{(-a+√(a^2+8g(z+L) ) d^2 } dz$$
Pertanyaan: a. Hitunglah waktu yang dibutuhkan infus untuk diganti! (Ketika z = 5 cm). Gunakan semua metode integrasi untuk soal ini!


In [ ]:
import math

def fungsi(z):
    return 1 / (-5852.16 + math.sqrt(5852.16**2 + 7848 * (z + 91.44)))

def integral(z):
    return (5852.16 * math.log(abs(math.sqrt(196200 * z + 25 * (5852.16**2 + 17940528 / 25)) - 5 * 5852.16)) +
            math.sqrt(7848 * z + 5852.16**2 + 17940528 / 25)) / 3924

a = 5
b = 50
h = 0.001
n = round((b - a) / h)

t1 = integral(a)
t2 = integral(b)

exact = math.fabs(80000 * (t1 - t2))
print('Integral exact    = ', exact)
print('----------------------------------------------------------')

# Metode Trapezoid
fa = fungsi(a)
fb = fungsi(b)
jum = 0

while a < (b - h):
    a = a + h
    fa = fungsi(a)
    jum = jum + fa

hasil_trapezoid = 80000 * h / 2 * (fa + 2 * jum + fb)
error_trapezoid = math.fabs(exact - hasil_trapezoid)

print('Integral Trapezoid  = ', hasil_trapezoid)
print('Error Trapezoid = ', error_trapezoid)
print('----------------------------------------------------------')

# Metode Simpson 1/3
a = 5
fa = fungsi(a)
n = 0
ganjil = 0
genap = 0

while a < (b - h):
    a = a + h
    n = n + 1
    if n % 2 != 0:
        ganjil = ganjil + fungsi(a)
    else:
        genap = genap + fungsi(a)

hasil_simpson = 80000 * h / 3 * (fa + 4 * ganjil + 2 * genap + fb)
error_simpson = math.fabs(exact - hasil_simpson)
print('Integral Simpson  = ', hasil_simpson)
print('Error Simpson = ', error_simpson)


Integral exact    =  45995.85330896104
----------------------------------------------------------
Integral Trapezoid  =  45996.506817745496
Error Trapezoid =  0.6535087844531517
----------------------------------------------------------
Integral Simpson  =  45996.420188377466
Error Simpson =  0.5668794164230349


Dari hasil yang diperoleh, dapat disimpulkan bahwa nilai eksak untuk waktu yang dibutuhkan oleh larutan garam dalam kantong infus, mulai dari ketinggian 50 cm hingga mencapai 5 cm, adalah sekitar 45995.85330896104 detik atau setara dengan kurang lebih 13 jam. Ketika menggunakan metode trapesium, hasil perhitungan menunjukkan angka 45996.506817745496 dengan nilai error sebesar 0.6535087844531517. Sementara itu, metode Simpson 1/3 memberikan hasil sekitar 45996.420188377466 dengan nilai error sebesar 0.5668794164230349. Jika dilakukan perbandingan antara kedua metode, terlihat bahwa metode Simpson 1/3 memiliki nilai error yang lebih kecil dibandingkan dengan metode trapesium. Dengan demikian, metode Simpson 1/3 dapat dianggap lebih akurat dalam mencari jawaban untuk permasalahan tugas kedua.

# Task 3
**Siklisasi polimer (teori Jacobson- Stockmayer)**
Rantai polimer mengambil sampel sejumlah besar orientasi yang berbeda dalam ruang dan waktu. Pergerakan rantai polimer diatur oleh proses stokastik (probabilistik). Terkadang, kedua ujung rantai polimer linier dapat saling mendekat dalam jarak yang reaktif. Jika ikatan terbentuk di antara kedua ujung polimer, reaksi tersebut disebut sebagai siklisasi. Dengan mempelajari probabilitas terjadinya hal ini, seseorang dapat memperkirakan laju di mana rantai linier diubah menjadi rantai melingkar. Pertimbangkan polimer linier dengan N tautan. Probabilitas bahwa kedua ujung rantai berada dalam jarak ikatan b satu sama lain diberikan oleh integral berikut:
$$[\frac{3}{2πNb^2}]^{3⁄2} ∫_0^bexp(\frac{-3r^2}{2Nb^2} ) 4πr^2 dr$$
Jika N = 20 mata rantai dan b = 1, hitunglah probabilitas ujung-ujung rantai berada dalam jarak b satu sama lain. Gunakan aturan trapesium komposit dua segmen dan kemudian aturan trapesium komposit empat segmen. Gunakan kedua pendekatan ini untuk mengekstrapolasi solusi yang lebih akurat dengan menghilangkan suku kesalahan O(h2), di mana h adalah lebar panel.



In [ ]:
import math
from sympy import *

N = 20
b = 1
a = 0

def fungsi(r):
    return (3 / (2 * math.pi * N * b**2))**(3 / 2) * exp((-3 * r**2) / (2 * N * b**2)) * 4 * math.pi * r**2

def integral(r):
    return -0.309019361618552 * r * exp(-3 * r**2 / 40) + 0.103006453872851 * math.sqrt(30) * math.sqrt(
        math.pi) * erf(math.sqrt(30) * r / 20)

t1 = integral(a)
t2 = integral(b)

exact = (t2 - t1)
print('Integral exact  = ', exact)
print('----------------------------------------------------')

# Metode Trapezoid
fa = fungsi(a)
fb = fungsi(b)

j = 0
n = [2, 4]

for i in range(0, len(n)):
    a = 0
    print('TRAPEZOID', n[i], 'SEGMEN')
    h = (b - a) / n[i]
    while a < (b - h):
        a = a + h
        fx = fungsi(a)
        j = j + fx
        s = (0.5 * h) * (fa + 2 * j + fb)
        er = math.fabs(s - exact)

    print('Hasil Metode Trapezoid  = ', round(s, 5))
    print('Error = ', round(er, 5))
    print('Lebar segmen (h)  = ', h)
    print('----------------------------------------------------\n')

# Metode Simpson 1/3
ganjil = 0
genap = 0
a = 0
h = 0.25
n = 0

while a < (b - h):
    a = a + h
    n = n + 1
    m = n % 2
    if m == 1:
        fx = fungsi(a)
        ganjil = ganjil + fx
    elif m == 0:
        fx = fungsi(a)
        genap = genap + fx

hasil = (h / 3) * (fa + 4 * ganjil + 2 * genap + fb)
error = math.fabs(hasil - exact)

print('METODE SIMPSON 1/3')
print('Hasil Metode Simpson 1/3 = ', round(hasil, 5))
print('Error = ', round(error, 5))
print('Lebar Segmen (h)  = ', h)


Integral exact  =  0.0147739418056433
----------------------------------------------------
TRAPEZOID 2 SEGMEN
Hasil Metode Trapezoid  =  0.01644
Error =  0.00166
Lebar segmen (h)  =  0.5
----------------------------------------------------

TRAPEZOID 4 SEGMEN
Hasil Metode Trapezoid  =  0.01803
Error =  0.00326
Lebar segmen (h)  =  0.25
----------------------------------------------------

METODE SIMPSON 1/3
Hasil Metode Simpson 1/3 =  0.01477
Error =  0.0
Lebar Segmen (h)  =  0.25


Berdasarkan hasil yang diperoleh, terlihat bahwa nilai eksak untuk menghitung peluang ujung rantai berada dalam jarak b = 1  satu sama lain, dengan jumlah mata rantai  N  sebanyak 20, adalah  r = 0.0147739418056433 . Selanjutnya, pendekatan dengan metode trapesium 2 segmen menghasilkan nilai  r = 0.01644  dengan nilai error sebesar 0.00166. Sementara itu, metode trapesium 4 segmen menghasilkan nilai  r = 0.01803  dengan nilai error 0.00326. Metode simpson 1/3 menunjukkan hasil  r = 0.01477  dengan nilai error 0.0.
Dalam metode trapesium, terlihat bahwa peningkatan jumlah segmen akan menyebabkan nilai  h  menjadi semakin kecil, menghasilkan error yang semakin kecil pula. Hasil yang diperoleh dari trapesium 4 segmen lebih baik dibandingkan dengan trapesium 2 segmen. Jika kedua metode dibandingkan, terlihat bahwa nilai error untuk metode simpson 1/3 lebih kecil dibandingkan dengan metode trapesium, menunjukkan bahwa metode simpson 1/3 lebih akurat dalam mencari jawaban untuk permasalahan yang diuraikan dalam tugas 3.
